pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,428 @@
+package symbolizer
+
+import (
+	"bytes"
+	"context"
+	"debug/elf"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"path/filepath"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+
+	googlev1 "github.com/grafana/pyroscope/api/gen/proto/go/google/v1"
+	"github.com/grafana/pyroscope/lidia"
+	"github.com/grafana/pyroscope/pkg/objstore"
+)
+
+type DebuginfodClient interface {
+	FetchDebuginfo(ctx context.Context, buildID string) (io.ReadCloser, error)
+}
+
+type Config struct {
+	Enabled                  bool                 `yaml:"enabled"`
+	DebuginfodURL            string               `yaml:"debuginfod_url"`
+	PersistentDebugInfoStore DebugInfoStoreConfig `yaml:"persistent_debuginfo_store"`
+}
+
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.BoolVar(&cfg.Enabled, "symbolizer.enabled", false, "Enable symbolization for unsymbolized profiles")
+	f.StringVar(&cfg.DebuginfodURL, "symbolizer.debuginfod-url", "https://debuginfod.elfutils.org", "URL of the debuginfod server")
+	cfg.PersistentDebugInfoStore.Storage.RegisterFlagsWithPrefix("symbolizer.persistent-debuginfo-store.storage.", f)
+}
+
+type Symbolizer struct {
+	logger  log.Logger
+	client  DebuginfodClient
+	store   DebugInfoStore
+	metrics *metrics
+}
+
+func New(logger log.Logger, cfg Config, reg prometheus.Registerer, bucket objstore.Bucket) (*Symbolizer, error) {
+	metrics := newMetrics(reg)
+
+	store := NewObjstoreDebugInfoStore(bucket)
+
+	client, err := NewDebuginfodClient(logger, cfg.DebuginfodURL, metrics)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Symbolizer{
+		logger:  logger,
+		client:  client,
+		store:   store,
+		metrics: metrics,
+	}, nil
+}
+
+func (s *Symbolizer) SymbolizePprof(ctx context.Context, profile *googlev1.Profile) error {
+	start := time.Now()
+	status := statusSuccess
+	defer func() {
+		s.metrics.profileSymbolization.WithLabelValues(status).Observe(time.Since(start).Seconds())
+	}()
+
+	mappingsToSymbolize := make(map[uint64]bool)
+	for i, mapping := range profile.Mapping {
+		if mapping.HasFunctions {
+			continue
+		}
+		mappingsToSymbolize[uint64(i+1)] = true
+	}
+	if len(mappingsToSymbolize) == 0 {
+		return nil
+	}
+
+	locationsByMapping, err := s.groupLocationsByMapping(profile, mappingsToSymbolize)
+	if err != nil {
+		return fmt.Errorf("grouping locations by mapping: %w", err)
+	}
+
+	stringMap := make(map[string]int64, len(profile.StringTable))
+	for i, str := range profile.StringTable {
+		stringMap[str] = int64(i)
+	}
+
+	funcMap := make(map[funcKey]uint64)
+	maxFuncID := uint64(len(profile.Function) + 1)
```

Comment:
@korniltsev I got a panic index out of bounds, so not sure we can rely on function IDs to be sequential (maybe something else is wrong?). For now I'm reverting the change back to:
```go
funcMap := make(map[funcKey]uint64)
maxFuncID := uint64(0)

for _, fn := range profile.Function {
    if fn.Id > maxFuncID {
	maxFuncID = fn.Id
    }
    funcMap[funcKey{fn.Name, fn.Filename}] = fn.Id
}
```

Reply:
can you reproduce? I suggest we understand / reproduce why it paniced before reverting. And maybe have a testcase.